File: sample/cpp/sample_img_face_attribute.cpp

#include "tdl_model_factory.hpp"

int main(int argc, char **argv) {
  if (argc != 4) {
    printf("Usage: %s <model_id_name> <model_dir> <image_path>\n", argv[0]);
    return -1;
  }
  std::string model_id_name = argv[1];
  std::string model_dir = argv[2];
  std::string image_path = argv[3];

  std::shared_ptr<BaseImage> image = ImageFactory::readImage(image_path);
  if (!image) {
    printf("Failed to create image\n");
    return -1;
  }

  TDLModelFactory &model_factory = TDLModelFactory::getInstance();
  model_factory.loadModelConfig();
  model_factory.setModelDir(model_dir);
  std::shared_ptr<BaseModel> model = model_factory.getModel(model_id_name);
  if (!model) {
    printf("Failed to create model\n");
    return -1;
  }

  std::vector<std::shared_ptr<ModelOutputInfo>> out_datas;
  std::vector<std::shared_ptr<BaseImage>> input_images = {image};
  model->inference(input_images, out_datas);
  const std::array<std::string, 7> id_to_emotion = {
      "Anger", "Disgust", "Fear", "Happy", "Neutral", "Sad", "Surprise"};
  for (size_t i = 0; i < out_datas.size(); i++) {
    if (out_datas[i]->getType() != ModelOutputType::CLS_ATTRIBUTE) {
      printf("out_datas[%ld] is not ModelOutputType::CLS_ATTRIBUTE\n", i);
      continue;
    }
    std::shared_ptr<ModelAttributeInfo> face_meta =
        std::static_pointer_cast<ModelAttributeInfo>(out_datas[i]);
    if (model_id_name == "CLS_ATTRIBUTE_GENDER_AGE_GLASS_MASK") {
      printf(
          "gender score:%f,age score:%f,glass score:%f,mask score:%f\n",
          face_meta->attributes
              [TDLObjectAttributeType::OBJECT_ATTRIBUTE_HUMAN_GENDER],
          face_meta
              ->attributes[TDLObjectAttributeType::OBJECT_ATTRIBUTE_HUMAN_AGE],
          face_meta->attributes
              [TDLObjectAttributeType::OBJECT_ATTRIBUTE_HUMAN_GLASSES],
          face_meta->attributes
              [TDLObjectAttributeType::OBJECT_ATTRIBUTE_HUMAN_MASK]);
      if (face_meta->attributes
              [TDLObjectAttributeType::OBJECT_ATTRIBUTE_HUMAN_GENDER] > 0.5) {
        printf("Gender:Male\n");
      } else {
        printf("Gender:Female\n");
      }
      printf("Age:%d\n",
             int(face_meta->attributes
                     [TDLObjectAttributeType::OBJECT_ATTRIBUTE_HUMAN_AGE] *
                 100));
      if (face_meta->attributes
              [TDLObjectAttributeType::OBJECT_ATTRIBUTE_HUMAN_GLASSES] > 0.5) {
        printf("Glass:Yes\n");
      } else {
        printf("Glass:No\n");
      }
      if (face_meta->attributes
              [TDLObjectAttributeType::OBJECT_ATTRIBUTE_HUMAN_MASK] > 0.5) {
        printf("Mask:Yes\n");
      } else {
        printf("Mask:No\n");
      }
    } else if (model_id_name == "CLS_ATTRIBUTE_GENDER_AGE_GLASS") {
      printf(
          "gender score:%f,age score:%f,glass score:%f\n",
          face_meta->attributes
              [TDLObjectAttributeType::OBJECT_ATTRIBUTE_HUMAN_GENDER],
          face_meta
              ->attributes[TDLObjectAttributeType::OBJECT_ATTRIBUTE_HUMAN_AGE],
          face_meta->attributes
              [TDLObjectAttributeType::OBJECT_ATTRIBUTE_HUMAN_GLASSES]);
      if (face_meta->attributes
              [TDLObjectAttributeType::OBJECT_ATTRIBUTE_HUMAN_GENDER] > 0.5) {
        printf("Gender:Male\n");
      } else {
        printf("Gender:Female\n");
      }
      printf("Age:%d\n",
             int(face_meta->attributes
                     [TDLObjectAttributeType::OBJECT_ATTRIBUTE_HUMAN_AGE] *
                 100));
      if (face_meta->attributes
              [TDLObjectAttributeType::OBJECT_ATTRIBUTE_HUMAN_GLASSES] > 0.5) {
        printf("Glass:Yes\n");
      } else {
        printf("Glass:No\n");
      }
    } else if (model_id_name == "CLS_ATTRIBUTE_GENDER_AGE_GLASS_EMOTION") {
      if (face_meta->attributes
              [TDLObjectAttributeType::OBJECT_ATTRIBUTE_HUMAN_GENDER] > 0.5) {
        printf("Gender:Male\n");
      } else {
        printf("Gender:Female\n");
      }
      printf("Age:%d\n",
             int(face_meta->attributes
                     [TDLObjectAttributeType::OBJECT_ATTRIBUTE_HUMAN_AGE] *
                 100));
      if (face_meta->attributes
              [TDLObjectAttributeType::OBJECT_ATTRIBUTE_HUMAN_GLASSES] > 0.5) {
        printf("Glass:Yes\n");
      } else {
        printf("Glass:No\n");
      }
      printf(
          "Emotion:%s\n",
          id_to_emotion
              [int(face_meta->attributes[TDLObjectAttributeType::
                                             OBJECT_ATTRIBUTE_HUMAN_EMOTION])]
                  .c_str());
    } else {
      printf("Not supported model id: %s\n", argv[1]);
      return -1;
    }
  }

  return 0;
}

File: sample/python/sample_img_face_attribute.py
import sys
import os
from tdl import nn, image

if __name__ == "__main__":
    if len(sys.argv) != 4:
        print("Usage: python sample_img_face_attribute.py <model_id_name> <model_path> <image_path>")
        sys.exit(1)
    model_id_name = sys.argv[1]
    model_path = sys.argv[2]
    img_path = sys.argv[3]

    model_id = getattr(nn.ModelType, model_id_name)
    face_attribute = nn.get_model(model_id, model_path)
    img = image.read(img_path)
    face_attribution = face_attribute.inference(img)
    result = face_attribution[0]

    if model_id_name == "CLS_ATTRIBUTE_GENDER_AGE_GLASS_MASK":
        print("gender_score:", result["gender_score"])
        print("is_male:", result["is_male"])
        print("age_score:", result["age_score"])
        print("age:", result["age"])
        print("glass_score:", result["glass_score"])
        print("is_wearing_glass:", result["is_wearing_glass"])
        print("mask_score:", result["mask_score"])
        print("is_wearing_mask:", result["is_wearing_mask"])
    elif model_id_name == "CLS_ATTRIBUTE_GENDER_AGE_GLASS":
        print("gender_score:", result["gender_score"])
        print("is_male:", result["is_male"])
        print("age_score:", result["age_score"])
        print("age:", result["age"])
        print("glass_score:", result["glass_score"])
        print("is_wearing_glass:", result["is_wearing_glass"])
    elif model_id_name == "CLS_ATTRIBUTE_GENDER_AGE_GLASS_EMOTION":
        print("gender_score:", result["gender_score"])
        print("is_male:", result["is_male"])
        print("age_score:", result["age_score"])
        print("age:", result["age"])
        print("glass_score:", result["glass_score"])
        print("is_wearing_glass:", result["is_wearing_glass"])
        print("emotion_score:", result["emotion_score"])
        print("emotion:", result["emotion"])
    else:
        print("Unknown model type:", model_id_name)

# input: python sample_img_face_attribute.py <model_id_name> <model_path> <image_path>
# output: gender_score: <value> is_male: <value> age_score: <value> age: <value> glass_score: <value> is_wearing_glass: <value> mask_score: <value> is_wearing_mask: <value>